ambry-cloud/src/main/java/com/github/ambry/cloud/azure/CosmosChangeFeedBasedReplicationFeed.java

```diff
@@ -169,49 +170,64 @@ public CosmosChangeFeedBasedReplicationFeed(CosmosDataAccessor cosmosDataAccesso
+      int index = cosmosChangeFeedFindToken.getIndex();
+      ChangeFeedCacheEntry changeFeedCacheEntry = changeFeedCache.get(cosmosChangeFeedFindToken.getCacheSessionId());
+      boolean cacheHit = true;
+      if (changeFeedCacheEntry == null || !isCacheValid(partitionPath, cosmosChangeFeedFindToken,
+          changeFeedCacheEntry)) {
+        // the cache may not be valid. So we cannot use session id
+        azureMetrics.changeFeedCacheMissRate.mark();
+        cacheHit = false;
+        changeFeedCacheEntry = getNextChangeFeed(partitionPath, cosmosChangeFeedFindToken.getStartContinuationToken());
+        // invalidate the previous token's cache
+        changeFeedCache.remove(cosmosChangeFeedFindToken.getCacheSessionId());
+        index = 0;
+      }
 
-    long resultSize = 0;
+      long resultSize = 0;
 
-    List<CloudBlobMetadata> fetchedEntries = changeFeedCacheEntry.getFetchedEntries();
-    while (true) {
-      if (index < fetchedEntries.size()) {
-        if (resultSize + fetchedEntries.get(index).getSize() < maxTotalSizeOfEntries || resultSize == 0) {
-          nextEntries.add(fetchedEntries.get(index));
-          resultSize = resultSize + fetchedEntries.get(index).getSize();
-          index++;
+      List<CloudBlobMetadata> fetchedEntries = changeFeedCacheEntry.getFetchedEntries();
+      while (true) {
+        if (index < fetchedEntries.size()) {
+          if (cacheHit) {
+            azureMetrics.changeFeedCacheHitRate.mark();
+            cacheHit = false;
+          }
+          if (resultSize + fetchedEntries.get(index).getSize() < maxTotalSizeOfEntries || resultSize == 0) {
+            nextEntries.add(fetchedEntries.get(index));
+            resultSize = resultSize + fetchedEntries.get(index).getSize();
+            index++;
+          } else {
+            break;
+          }
         } else {
-          break;
-        }
-      } else {
-        // we can reuse the session id in this case, because we know that the cache ran out of new items.
-        changeFeedCacheEntry = getNextChangeFeed(partitionPath, changeFeedCacheEntry.getEndContinuationToken(),
-            changeFeedCacheEntry.getCacheSessionId());
-        fetchedEntries = changeFeedCacheEntry.getFetchedEntries();
-        if (fetchedEntries.isEmpty()) {
-          // this means that there are no new changes
-          break;
+          // we can reuse the session id in this case, because we know that the cache ran out of new items.
+          changeFeedCacheEntry = getNextChangeFeed(partitionPath, changeFeedCacheEntry.getEndContinuationToken(),
+              changeFeedCacheEntry.getCacheSessionId());
+          fetchedEntries = changeFeedCacheEntry.getFetchedEntries();
+          if (fetchedEntries.isEmpty()) {
+            // this means that there are no new changes
+            break;
+          } else {
+            azureMetrics.changeFeedCacheRefreshRate.mark();
+          }
+          index = 0;
```

Comment:
I probably misunderstand this but I am thinking if there is a case where `cacheHit` is true before entering the `while(true)` loop.  The first time code enters in while loop, `index 
 == fetchedEntries.size()` and jumps to `else` branch. It does fetch new changes and index is reset to 0. Then it goes back to beginning of while(true) and `azureMetrics.changeFeedCacheHitRate.mark();` will be called. I just want to confirm this is a possible case and updating `changeFeedCacheHitRate` is what we expect.

Reply:
First time code enters in while loop, and if index == fetchedEntries.size(), then we refresh the cache as you mentioned. In this case we don't want to say cacheHit just yet, because we haven't yet hit the cache. We will refresh the cache and mark it so.

Now once the refresh is successful, there can be either no results (fetchedEntries.isEmpty()), in which case we will come out of loop without marking cache hit (as expected because it isn't technically a cache hit). If there are entries returned  in cache refresh, then loop's next iteration will run, which will mark the cache hit. 

In short unless we are able to consume something from the cache, we don't mark a cache hit. Hope this makes sense.

Reply:
make sense